src/main/java/LRParser.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    public boolean hasNonterminalAfterDot (Production production) {
+        if (production.getNextSymbol() != null) {
+            return grammar.nonterminals.contains(production.getNextSymbol());
+        }
+        return false;
+    }
+
+    State goTo (State state, String symbol) {
+        Set<Production> productions = new HashSet<>();
+        for (Production production : state.getProductionSet()) {
+            String nextSymbol = production.getNextSymbol();
+            if (nextSymbol != null && nextSymbol.equals(symbol)) {
+                var copyProduction = new Production(production);
+                productions.add(copyProduction);
+                copyProduction.movePoint();
+            }
+        }
+        return closure(productions);
+    }
+
+    State closure (Set<Production> productions) {
+        State newState = new State();
+        newState.productionSet.addAll(productions);
+        Boolean isFinished = false;
+        while (!isFinished) {
+            isFinished = true;
+            List<Production> productionsToAdd = new ArrayList<>();
+            for (Production production : newState.productionSet) {
+                if (hasNonterminalAfterDot(production)) {
+                    String nonterminal = production.rhs.get(production.pointIndex);
+                    List<Production> productionList = grammar.printProductionsForGivenNonTerminal(nonterminal);
+                    for (Production production1 : productionList) {
+                        if (!newState.productionSet.contains(production1)) {
+                            productionsToAdd.add(production1);
+                            isFinished = false;
+                        }
+                    }
+                }
+            }
+            newState.productionSet.addAll(productionsToAdd);
+
+        }
+        return newState;
+    }
+
+    void computeCanonicalCollection () {
+        State initialState = new State();
+        initialState.productionSet.add(grammar.enhanceGrammar());
+        initialState = closure(initialState.productionSet);
+        canonicalCollection.add(initialState);
+        boolean isFinished = false;
+        List<String> symbols = new ArrayList<>();
+        symbols.addAll(grammar.nonterminals);
+        symbols.addAll(grammar.terminals);
+        Integer lastIndex = 0;
+        while (!isFinished) {
+            isFinished = true;
+            for (int i = 0; i < canonicalCollection.size(); i++) {
+                State state = canonicalCollection.get(i);
+                for (String symbol : symbols) {
+                    State newState = goTo(state, symbol);
+                    if (!newState.productionSet.isEmpty() && !canonicalCollection.contains(newState)) {
+                        canonicalCollection.add(newState);
+                        isFinished = false;
+                        break;
+                    }
+                    if (!isFinished) {
+                        break;
```

Comment:
I don't think that all previous states should be checked again, there will be multiple repetitive gotos performed. Instead, I think that by skipping all break lines and the while loop it should also work without repeating too many goto operations, by just iterating through the continuously growing canonical collection and the symbols of each state.

Reply:
I tried iterating through the continuously growing canonical collection but I got a ConcurrentLock Exception because I was modifying the canonical collection while iterating through it so I had to save what needs to be added and actually add them at the end.